### DISCUSSION TITLE
Embedded Shape Drawing Support

### DISCUSSION THREAD

**Discussion by tedzards509** [lichess-org/flutter-chessground#31] [Announcements]
You wanted the users to be able to draw shapes by touching with one finger and moving the other. This is not possible using a gesture detector as far as I could tell, so a custom solution using listeners to touch events is necessary one way or another. Implementing this without Riverpod seems tedious, so I think the optimal solution would be to refactor it using callback functions provided by the app using flutter-chessground while providing a "default" GestureDetector to slot in which doesn't allow for drawing shapes natively.

When using Riverpod or at least callbacks, it is also possible to differentiate events by input device type, such as a stylus. This could allow for the users to draw shapes with a stylus while playing normally with fingers.

**Comment by tedzards509** [comment]
A Flutter project that implemented differentiation between stylus and finger input would be [this calendar app](https://github.com/AlexanderThiele/stift_flutter_app). Notably around [this](https://github.com/AlexanderThiele/stift_flutter_app/blob/53c0827f42081d6ce893d6942bbd38335d431c02/lib/components/calendar_table/interactive_paint_view.dart#L167) code section.

**Comment by veloce** [comment]
I would rather keep this project dependency free, and not use riverpod. Riverpod is good for a complex app, but for a package like this, where there isn't much state, basic state management using `StatefulWidget` is preferable.

**Comment by veloce** [comment]
Also I think this handled only in flutter-chessground, not an external app through callbacks. 

Not sure yet whether we should use a `RawGestureDetector` with a `MultiDragGestureRecognizer` or directly the low level `Listener`.

My guess is the `Listener` would be more appropriate.
